library/core/src/mem/mod.rs

```diff
@@ -80,7 +80,7 @@ pub use crate::intrinsics::transmute;
 ///
 /// # Relationship with `ManuallyDrop`
 ///
-/// While `mem::forget` can also be used to transfer *memory* ownership, doing so is error-prone.
+/// While `mem::forget` can also be used to *transfer* memory ownership, doing so is error-prone.
```

Comment:
The old emphasis was deliberate. The previous paragraph talks about transferring other resources, now it talks about memory.

Why would it makes sense to emphasize "transfer"?

Reply:
The specificity of this section to "memory" seems incidental. The way I read it, the primary point of this section is to advise erring in the direction of running drop 0 times (usually not UB) instead of running drop 2+ times (more often problematic). That advice, and the suitability of `ManuallyDrop` for this purpose, is not specific to Drop impls which involve memory ownership. It applies to file descriptors, lock guards, anything that might get transferred from one form to another and get simultaneously covered by 2 destructors at some point during that transfer.

The correct usage of `mem::forget` that I am familiar with in my own code is always in the form of **defusing** as opposed to **transferring**:

- https://github.com/dtolnay/cxx/blob/1.0.107/src/unwind.rs#L5-L28
- https://github.com/dtolnay/no-panic/blob/0.1.26/src/lib.rs#L271-L280

In the links, `ManuallyDrop` is simply not usable, and the reason is that no transfer is taking place, only defusal.

So the contrast being made isn't non-memory vs memory, it is non-transfer vs transfer, with the advice being for transfers to be expressed using `ManuallyDrop`.

Reply:
> the primary point of this section

This should say "one of the two points". The paragraphs related to provenance apply to memory. The paragraphs about double-drop apply generally to anything that is a transfer.

Reply:
The original intent of this paragraph was to say that `mem::forget` is fine for file descriptors and the like, but `ManuallyDrop` should be used for memory. One major motivation for that is UB due to aliasing model violations, which applies specifically only to memory. That's what the example in this section illustrates as well.

If the general opinion right now is to always use `ManuallyDrop` when possible and only use `forget` for these "defusing" situations, then a larger structural change to the docs is required I think? (E.g., the [example](https://doc.rust-lang.org/nightly/std/mem/fn.forget.html#examples) needs changing.)

Reply:
With this PR, we have an example that does transfer using `forget` (it literally speaks about something that was "previously transferred to code outside of Rust"), and then we have another section that emphasizes transfer -- in contrast to what? Both examples show transfer!